fog/ledger/enclave/impl/src/oblivious_utils.rs

```diff
@@ -0,0 +1,454 @@
+        &shard_key_image_search_result.key_image_result_code,
+    );
+
+    client_key_image_search_result.spent_at.cmov(
+        should_overwrite_key_image_search_result,
+        &shard_key_image_search_result.spent_at,
+    );
+
+    client_key_image_search_result.timestamp.cmov(
+        should_overwrite_key_image_search_result,
+        &shard_key_image_search_result.timestamp,
+    );
+
+    client_key_image_search_result.timestamp_result_code.cmov(
+        should_overwrite_key_image_search_result,
+        &shard_key_image_search_result.timestamp_result_code,
+    );
+}
+
+fn should_overwrite_key_image_search_result(
+    client_key_image_search_result: &KeyImageResult,
+    shard_key_image_search_result: &KeyImageResult,
+) -> Choice {
+    let client_key_image: &[u8] = client_key_image_search_result.key_image.as_ref();
+    let shard_key_image: &[u8] = shard_key_image_search_result.key_image.as_ref();
+    let do_key_images_match = client_key_image.ct_eq(shard_key_image);
+
+    let client_key_image_search_result_code = client_key_image_search_result.key_image_result_code;
+    let shard_key_image_search_result_code = shard_key_image_search_result.key_image_result_code;
+
+    let client_code_is_not_spent: Choice =
+        client_key_image_search_result_code.ct_eq(&(KeyImageResultCode::NotSpent as u32));
+
+    let shard_code_is_spent: Choice =
+        shard_key_image_search_result_code.ct_eq(&(KeyImageResultCode::Spent as u32));
+    let shard_code_is_error: Choice =
+        shard_key_image_search_result_code.ct_eq(&(KeyImageResultCode::KeyImageError as u32));
+
+    //   We make the same query to several shards and get several responses, and
+    // this logic determines how we fill the one client response.
+    //   At a high level, we want to prioritize "spent" responses.
+    // Error responses are "retriable" errors that the client will retry
+    // after a backoff. The "not spent" response is the default response and
+    // gets overwritten by any other response.
+    // spent > error > not spent
+    do_key_images_match
+           // Always write a Found code
+        & (shard_code_is_spent
+            // Write an error code IFF the client code is NotFound.
+            | ((shard_code_is_error) & client_code_is_not_spent))
+}
+
+#[cfg(test)]
+mod tests {
+    extern crate std;
+
+    use super::*;
+    use std::vec;
+
+    #[test]
+    fn should_overwrite_tests() {
+        let test_cases = vec![
+            // Images don't match
+            (
+                123456,
+                1,
+                123456,
+                TimestampResultCode::TimestampFound,
+                KeyImageResultCode::NotSpent,
+                654321,
+                1,
+                123456,
+                TimestampResultCode::TimestampFound,
+                KeyImageResultCode::Spent,
+                false,
+                "key images don't match, but overwritten anyway!",
+            ),
```

Comment:
Nit: Make a struct that holds these values. It's difficult to understand what these variables mean without looking at the for loop below. This would also simplify the for loop and allow you to access fields rather than writing variable names in the tuple.

Reply:
💯